apis/core.oam.dev/v1beta1/resourcetracker_types.go

```diff
@@ -69,9 +70,61 @@ const (
 
 // ResourceTrackerSpec define the spec of resourceTracker
 type ResourceTrackerSpec struct {
-	Type                  ResourceTrackerType `json:"type,omitempty"`
-	ApplicationGeneration int64               `json:"applicationGeneration"`
-	ManagedResources      []ManagedResource   `json:"managedResources,omitempty"`
+	Type                  ResourceTrackerType        `json:"type,omitempty"`
+	ApplicationGeneration int64                      `json:"applicationGeneration"`
+	ManagedResources      []ManagedResource          `json:"managedResources,omitempty"`
+	Compression           ResourceTrackerCompression `json:"compression,omitempty"`
+}
+
+// ResourceTrackerCompression the compression for ResourceTracker ManagedResources
+type ResourceTrackerCompression struct {
+	Type compression.Type `json:"type,omitempty"`
+	Data string           `json:"data,omitempty"`
+}
+
+// MarshalJSON .
+func (in *ResourceTrackerSpec) MarshalJSON() ([]byte, error) {
+	type Alias ResourceTrackerSpec
+	tmp := &struct{ *Alias }{}
+	switch in.Compression.Type {
+	case compression.Uncompressed:
+		tmp.Alias = (*Alias)(in)
+	case compression.Gzip:
+		cpy := in.DeepCopy()
+		data, err := compression.GzipObjectToString(in.ManagedResources)
+		if err != nil {
+			return nil, err
+		}
+		cpy.ManagedResources = nil
+		cpy.Compression.Data = data
+		tmp.Alias = (*Alias)(cpy)
+	default:
+		return nil, compression.NewUnsupportedCompressionTypeError(string(in.Compression.Type))
+	}
+	return json.Marshal(tmp.Alias)
+}
+
+// UnmarshalJSON .
```

Comment:
```suggestion
// UnmarshalJSON wil unmarshal resource tracker spec according to the compression type, if feature enabled, it will unmarshal data from compression data. Note: this is not actually json unmarhsal but re-use the framework function.
```

Reply:
Done